### PR TITLE
feat(datepicker): add keyboard support to calendar

### PR DESCRIPTION
**Pull request by mmalerba**
temporary demo: https://mmalerba-demo1.firebaseapp.com/datepicker

### CODE REVIEW

File: src/lib/datepicker/calendar.ts [hunk 1]
```diff
@@ -97,57 +107,181 @@ export class MdCalendar implements AfterContentInit {
...
+      newDate = new SimpleDate(newDate.year, newDate.month, 0);
+    }
+
+    return newDate;
+  }
+
+  /**
+   * Adds the given number of months to the date. Months are counted as if flipping 12 pages for
+   * each year on a calendar and then finding the closest date in the new month. For example when
+   * adding 1 year to Feb 29 2016 the resulting date will be Feb 28 2017.
```
Comment (kara) [inline src/lib/datepicker/calendar.ts]: Ultra nit: add comma after 'Feb 29, 2016' (same for line 241)? Love the detailed comment though

File: src/lib/datepicker/calendar.ts [hunk 2]
```diff
@@ -97,57 +107,181 @@ export class MdCalendar implements AfterContentInit {
...
+    return date.add({'days': days});
+  }
+
+  /**
+   * Adds the given number of months to the date. Months are counted as if flipping pages on a
+   * calendar and then finding the closest date in the new month. For example when adding 1 month to
+   * Jan 31 2017 the resulting date will be Feb 28 2017.
+   */
+  private _addCalendarMonths(date: SimpleDate, months: number): SimpleDate {
+    let newDate = date.add({'months': months});
```
Comment (kara) [inline src/lib/datepicker/calendar.ts]: You can shorten this to

```ts
let newDate = date.add({months})
```

(same on line 235)

File: src/lib/datepicker/calendar.ts [hunk 3]
```diff
@@ -97,57 +107,181 @@ export class MdCalendar implements AfterContentInit {
...
   /** Whether the two dates represent the same view in the current view mode (month or year). */
-  private _isSameView(date1: SimpleDate, date2: SimpleDate) {
+  private _isSameView(date1: SimpleDate, date2: SimpleDate): boolean {
     return this._monthView ?
         date1.year == date2.year && date1.month == date2.month :
         date1.year == date2.year;
   }
+
+  /** Handles keydown events on the calendar body. */
+  _handleCalendarBodyKeydown(event: KeyboardEvent): void {
```
Comment (kara) [inline src/lib/datepicker/calendar.ts]: Considered breaking this logic out into its own mini-service? 

Also, have you thought about separating into two functions -one for month view and one for year view? Not sure if that will actually shorten anything, but it may be easier to scan it without the conditionals.

File: src/lib/datepicker/calendar.spec.ts [hunk 4]
```diff
@@ -231,6 +246,347 @@ describe('MdCalendar', () => {
...
+        calendarBodyEl = calendarElement.querySelector('.mat-calendar-body') as HTMLElement;
+        expect(calendarBodyEl).not.toBeNull();
+
+        dispatchFakeEvent(calendarBodyEl, 'focus');
+        fixture.detectChanges();
+      });
+
+      it('should not allow selection of disabled date in month view', () => {
+        expect(calendarInstance._monthView).toBe(true);
+        expect(calendarInstance._activeDate).toEqual(new SimpleDate(2017, 0, 1));
```
Comment (kara) [inline src/lib/datepicker/calendar.spec.ts]: Curious about this behavior. It seems like we allow people to focus the disabled dates, but not select them, whereas AngularJS Material doesn't allow highlighting of disabled dates at all. What's the a11y best practice for this?

File: src/lib/datepicker/calendar.spec.ts [hunk 5]
```diff
@@ -231,6 +246,347 @@ describe('MdCalendar', () => {
...
+        dispatchKeyboardEvent(calendarBodyEl, 'keydown', ENTER);
+        fixture.detectChanges();
+
+        expect(calendarInstance._monthView).toBe(true);
+        expect(testComponent.selected).toBeNull();
+      });
+    });
+  });
+
+  describe('a11y', () => {
```
Comment (kara) [inline src/lib/datepicker/calendar.spec.ts]: Nit: Two describe blocks both named `a11y`?

Comment (mmalerba) [inline src/lib/datepicker/calendar.spec.ts]: One of them is `MdCalendar.calendar with date filter.a11y` the other is `MdCalendar.a11y` (though it should be `MdCalendar.standard calendar.a11y` so making that change).

File: src/lib/datepicker/calendar.spec.ts (hunk 4 above)
Comment (mmalerba) [inline src/lib/datepicker/calendar.spec.ts]: Yeah this was more of a practical decision, since someone could theoretically create a calendar with no maxDate but a date filter that disables all dates > 1/1/17. In this case if the user tried to arrow past 1/1/17 everything would just lock up as it searched for the next enabled date. I put a TODO and I'm curious to see how AngularJS Material handled this

File: src/lib/datepicker/calendar.spec.ts (hunk 5 above)
Comment (kara) [inline src/lib/datepicker/calendar.spec.ts]: Aha I see. The diff was a bit misleading.

File: src/lib/datepicker/calendar.ts (hunk 3 above)
Comment (mmalerba) [inline src/lib/datepicker/calendar.ts]: I'm not sure what we'd gain by making it a service, I split out 2 separate functions for month view and year view though

**Comment by mmalerba** [comment]
Yeah focus state when you open the calendar is still kind of wonky, I'll address that in a different PR. Addressed other comments, PTAL.

File: src/lib/datepicker/calendar.ts (hunk 3 above)
Comment (kara) [inline src/lib/datepicker/calendar.ts]: Just thought the file was getting pretty long, but if you don't think so, doesn't matter too much.

File: src/lib/datepicker/calendar.ts [hunk 6]
```diff
@@ -97,57 +107,208 @@ export class MdCalendar implements AfterContentInit {
...
+
+  /** Adds the given number of days to the date. */
+  private _addCalendarDays(date: SimpleDate, days: number): SimpleDate {
+    return date.add({days});
+  }
+
+  /**
+   * Adds the given number of months to the date. Months are counted as if flipping pages on a
+   * calendar and then finding the closest date in the new month. For example when adding 1 month to
+   * Jan 31, 2017 the resulting date will be Feb 28, 2017.
```
Comment (kara) [inline src/lib/datepicker/calendar.ts]: Haha, sorry for being unclear. I meant a comma after "2017", not in the date.

```
For example, when adding 1 month to Jan 31, 2017, the resulting date will be ...
```
